src/java/org/apache/cassandra/db/compaction/TimeOrderedKeyCompactionStrategy.java

```diff
@@ -0,0 +1,950 @@
+        final TimeWindow timeWindowMs;
+
+        public SSTableStats(SSTableReader sstable)
+        {
+            StatsMetadata meta = sstable.getSSTableMetadata();
+            this.sstable = sstable;
+            this.tombstoneCounts = new TombstoneCounts();
+            this.tombstoneCounts.partitionTombstones += meta.partitionTombstones;
+            this.tombstoneCounts.rowTombstones += meta.rowTombstones;
+            this.tombstoneCounts.rangeTombstones += meta.rangeTombstones;
+
+            ICardinality cardinality = null;
+            try
+            {
+                cardinality = SSTableReader.getCardinality(sstable);
+            }
+            catch (IOException e)
+            {
+            }
+            this.cardinality = cardinality;
+
+            if (this.cardinality != null)
+            {
+                this.keyCount = cardinality.cardinality();
+            }
+            else
+            {
+                this.keyCount = sstable.estimatedKeys();
+            }
+            this.rowCount = sstable.getTotalRows();
+            this.timeWindowMs = new TimeWindow(meta.minKey, (int) (meta.maxKey - meta.minKey));
+        }
+
+        @Override
+        public boolean equals(Object o)
+        {
+            if (this == o) return true;
+            if (o == null || getClass() != o.getClass()) return false;
+            SSTableStats that = (SSTableStats) o;
+            return sstable.equals(that.sstable);
+        }
+
+        @Override
+        public int hashCode()
+        {
+            return Objects.hash(sstable);
+        }
+
+        @Override
+        public String toString()
+        {
+            return "SSTableStats{" +
+                    "sstable=" + sstable +
+                    ", cardinality=" + cardinality +
+                    ", keyCount=" + keyCount +
+                    ", rowCount=" + rowCount +
+                    ", tombstoneCounts=" + tombstoneCounts +
+                    ", timeRange=" + timeWindowMs + " ms" +
+                    '}';
+        }
+    }
+
+    static class OverallStats
+    {
+        long keyCount = 0;
+        long rowCount = 0;
+        TombstoneCounts tombstoneCounts = new TombstoneCounts();
+        long onDiskLength = 0;
+
+        @Override
+        public String toString()
+        {
+            return "OverallStats{" +
+                    "keyCount=" + keyCount +
+                    ", rowCount=" + rowCount +
+                    ", tombstoneCounts=" + tombstoneCounts +
+                    ", onDiskLength=" + onDiskLength +
+                    '}';
+        }
+    }
+
+    private static void collect(OverallStats os, SSTableStats ss)
+    {
+        if (ss.sstable.getSSTableLevel() == SSTable.DATA_SSTABLE_LVL)
+        {
+            os.keyCount += ss.keyCount;
+            os.rowCount += ss.rowCount;
+            os.onDiskLength += ss.sstable.onDiskLength();
+        }
+        else
+        {
+            os.tombstoneCounts.add(ss.tombstoneCounts);
```

Comment:
Why do we exclude tombstone tables from onDiskLength?

Reply:
On first thought it is just easier to assume that tombstones have negligible data size.
In my testing, 240K tombstones resulted in few MB of data.

Reply:
Makes sense.